test/e2e/harness.ts

```diff
@@ -0,0 +1,34 @@
+import { Page } from "@recordreplay/playwright";
+
+declare global {
+  const app: any;
+}
+
+export default class TestHarness {
+  constructor(private readonly page: Page) {}
+
+  async start() {
+    await this.page.waitForFunction(() => window.app?.actions);
+    this.page.evaluate(() => window.app.actions.setViewMode("dev"));
+    await this.page.waitForFunction(() => {
+      const loadedRegions = app.selectors.getLoadedRegions();
+      const fullyLoaded =
+        loadedRegions &&
+        loadedRegions.loading.length > 0 &&
+        loadedRegions.loaded.length > 0 &&
+        loadedRegions.loading[0].end.point === loadedRegions.loaded[0].end.point;
+      return (
+        fullyLoaded &&
+        app.selectors.getViewMode() == "dev" &&
+        document.querySelector(".webconsole-app")
+      );
+    });
+  }
+
+  waitForMessage(text: string) {
+    return this.page.waitForFunction(text => {
+      const messages = document.querySelectorAll(".webconsole-output .message");
+      return [...messages].some(message => (message as HTMLElement).innerText.includes(text));
```

Comment:
I'm kind of surprised that there isn't a `expect(this.page).toHaveSelector` or something we could use here?

Reply:
`expect` comes from jest, `page` comes from playwright so I'd be surprised if that exists.

Reply:
https://playwright.dev/docs/test-assertions#locator-assertions-to-have-text

Reply:
Oh, nice. However, it doesn't work for some reason: `page.locator()` doesn't exist according to typescript. So I tried with `(page as any).locator()`, which works, but then `expect(locator).toHaveText(...)` throws "TypeError: expect(...).toHaveText is not a function". I'll investigate this in a follow-up.